src/garage/sampler/rl2_sampler.py

```diff
@@ -0,0 +1,224 @@
+
+    Args:
+        algo (garage.np.algos.RLAlgorithm): An algorithm instance.
+        env (List[garage.envs.GarageEnv]): Environements to sample from.
+        meta_batch_size (int): Meta batch size for sampling. If it is
+            larger than n_envs, it must be a multiple of n_envs so it can be
+            evenly distributed among environments.
+        n_envs (int): Number of environment instances for sampling. It it is
+            larger than meta_batch_size, it must be a multiple of
+            meta_batch_size so batch can be evenly distributed among
+            environments.
+
+    """
+
+    def __init__(self, algo, env, meta_batch_size, n_envs=None):
+        if n_envs is None:
+            n_envs = singleton_pool.n_parallel * 4
+        self.algo = algo
+        self.envs = env
+
+        self._n_envs = n_envs
+        self._meta_batch_size = meta_batch_size
+        self._vec_env = None
+        self._envs_per_worker = None
+        self._vec_envs_indices = None
+
+    def start_worker(self):
+        """Start workers.
+
+        Raises:
+            ValueError: If meta_batch_size > n_envs and meta_batch_size is not
+                a multiple of n_envs, or if n_envs > meta_batch_size and n_envs
+                is not a multiple of meta_batch_size.
+
+        """
+        if self._meta_batch_size > self._n_envs:
+            if self._meta_batch_size % self._n_envs != 0:
+                raise ValueError(
+                    'meta_batch_size must be a multiple of n_envs')
+            self._envs_per_worker = 1
+            self._vec_envs_indices = np.split(np.arange(self._meta_batch_size),
+                                              self._n_envs)
+        if self._n_envs >= self._meta_batch_size:
+            if self._n_envs % self._meta_batch_size != 0:
+                raise ValueError(
+                    'n_envs must be a multiple of meta_batch_size')
+            self._envs_per_worker = self._n_envs // self._meta_batch_size
+            self._vec_envs_indices = [np.arange(self._meta_batch_size)]
+
+    def shutdown_worker(self):
+        """Shutdown workers."""
+        self._vec_env.close()
+
+    def setup_worker(self, env_indices):
+        """Setup workers.
+
+        Args:
+            env_indices (List[Int]): Indices of environments to be assigned
+                to workers for sampling.
+
+        """
+        vec_envs = []
+        for env_ind in env_indices:
+            vec_env = pickle.loads(pickle.dumps(self.envs[env_ind]))
+            vec_envs.extend([vec_env for env in range(self._envs_per_worker)])
+        seed0 = deterministic.get_seed()
+        if seed0 is not None:
+            for (i, e) in enumerate(vec_envs):
+                e.seed(seed0 + i)
+
+        self._vec_env = VecEnvExecutor(
+            envs=vec_envs, max_path_length=self.algo.max_path_length)
+
+    # pylint: disable=too-many-statements
+    def obtain_samples(self, itr, batch_size=None, whole_paths=True):
```

Comment:
I've read through this function three times, and I still don't understand why it's different from the other samplers.

Please write much more documentation about what's going on or split this function up.

Reply:
Good point. I will add more docstring. In summary, RL^2 needs to concatenate all paths for each environment/task, so we need to keep track of the source of each pat. Here we will have `paths[idx]` containing all paths for the `idx`-th environment/task.